s2i/builder.go

```diff
@@ -0,0 +1,115 @@
+package s2i
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	dockerClient "github.com/docker/docker/client"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/api/validation"
+	"github.com/openshift/source-to-image/pkg/build"
+	"github.com/openshift/source-to-image/pkg/build/strategies"
+	"github.com/openshift/source-to-image/pkg/scm/git"
+
+	fn "knative.dev/kn-plugin-func"
+	docker "knative.dev/kn-plugin-func/docker"
+)
+
+var (
+	// ErrRuntimeRequired indicates the required value of Function Runtime was not provided
+	ErrRuntimeRequired = errors.New("runtime is required to build")
+
+	// ErrRuntimeNotSupported indicates the given runtime is not (yet) supported
+	// by this builder.
+	ErrRuntimeNotSupported = errors.New("runtime not supported")
+)
+
+// DefaultBuilderImages for s2i builders indexed by Runtime Language
+var DefaultBuilderImages = map[string]string{
+	"node": "registry.access.redhat.com/ubi8/nodejs-16", // TODO: finalize choice and include version
+}
+
+// Builder of Functions using the s2i subsystem.
+type Builder struct {
+	verbose bool
+}
+
+// NewBuilder creates a new instance of a Builder with static defaults.
+func NewBuilder(verbose bool) *Builder {
+	return &Builder{verbose: verbose}
+}
+
+func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
+	// Ensure the Function has a builder specified
+	if f.Builder == "" {
+		f.Builder, err = defaultBuilderImage(f)
+		if err != nil {
+			return
+		}
+	}
+
+	client, endpoint, err := docker.NewClient(dockerClient.DefaultDockerHost)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	// TODO: this defaulting of endpoint should perhaps not be necessary?  The
+	// default is explicitly passed to NewClient, thus the returned endpoint
+	// should at least be that.
+	if endpoint == "" {
+		endpoint = dockerClient.DefaultDockerHost
+	}
+
+	// Build Config
+	cfg := &api.Config{}
+	cfg.Quiet = !b.verbose
+	cfg.Tag = f.Image
+	cfg.Source = &git.URL{URL: url.URL{Path: f.Root}, Type: git.URLTypeLocal}
+	cfg.BuilderImage = f.Builder
+	cfg.BuilderPullPolicy = api.DefaultBuilderPullPolicy
+	cfg.PreviousImagePullPolicy = api.DefaultPreviousImagePullPolicy
+	cfg.RuntimeImagePullPolicy = api.DefaultRuntimeImagePullPolicy
+	cfg.DockerConfig = &api.DockerConfig{
```

Comment:
There is `GetDefaultDockerConfig()` function that can default init this.

Reply:
Also is that `Endpoint` really used or only some validation is complaining? We are passing docker client so that should be used.

Reply:
`GetDefaultDockerConfig` seems to be a better choice, now that I know it exists.  I have replaced this manual config creation with that call, and therefore ignoring the endpoint returned from the docker client constructor.

Reply:
I think that the config is not used anyway. It could be used to construct an instance of `dockerClient.CommonAPIClient`, but we are passing an instance of the interface to the API so there is no need to use that I believe.